### PR TITLE
First stab at a whitelist for the xstream library.

### PR DESCRIPTION
**Pull request by FSchumacher**
`ant test` will run with the current whitelist, but I fear that most plugins would fail.

It would probably be cumbersome for users to search for the correct whitelist to setup JMeter for their beloved plugins. But what are the alternatives?

### CODE REVIEW

File: src/core/org/apache/jmeter/util/JMeterUtils.java [hunk 1]
```diff
@@ -1261,10 +1265,39 @@ public static final void refreshUI() {
...
+        for (TypePermission perm : Arrays.asList(NullPermission.NULL,
+                ArrayTypePermission.ARRAYS,
+                PrimitiveTypePermission.PRIMITIVES)) {
+            xstream.addPermission(perm);
+        }
+        for (Class<?> allowHierarchy : Arrays.asList(java.util.Collection.class,
+                org.apache.jmeter.testelement.TestElement.class,
+                org.apache.jorphan.collections.HashTree.class,
+                org.apache.jmeter.samplers.SampleSaveConfiguration.class)) {
+            xstream.allowTypeHierarchy(allowHierarchy);
```
Comment (pmouawad) [inline src/core/org/apache/jmeter/util/JMeterUtils.java]: Shouldn't we also allow subclasses of SampleResult ?

**Comment by pmouawad** [comment]
Hi @FSchumacher ,
Thanks for the PR.
Maybe we could add a interface that plugins could implement to register additional classes:

Too coarse and risky ?:
`SerializationRegistrator#registerClasses(XStream xstream)`

Other option:
`
Class[] SerializationRegistrator#getTypeHierarchies()
String[] SerializationRegistrator#getTypeByRegex()
String[] SerializationRegistrator#getTypeByWildcard()
`

Still considering what this could break, I am against including it in 3.3 (at least) and also it needs through testing by 3rd parties to have further feedback on what it breaks.
It looks to me like a Plugins Geddon that could hurt hardly our software.

File: src/core/org/apache/jmeter/util/JMeterUtils.java (hunk 1 above)
Comment (FSchumacher) [inline src/core/org/apache/jmeter/util/JMeterUtils.java]: Done.

**Comment by FSchumacher** [comment]
Now both the whitelist should be checked first and if the class is **not** allowed a warning will be logged (as info).
To support dynamic additions by plugins, we could have a look at Emi's suggestion of a ServiceLoader.

**Comment by vlsi** [comment]
Note: this PR was automatically closed when JMeter development moved from SVN to Git (main branch was renamed from `trunk` to `master`, thus GitHub closed the PR).
If you want to revive the PR, please update your fork as per https://github.com/vlsi/jmeter-git-cleanup/tree/v2.2.0/clean_fork , reopen the PR (unfortunately, it has to be a new PR).

If you have questions, please feel free to ask on JMeter dev list.

Sorry for the inconvenience

**Pull request closed without merging**